File: lab1/lab1a/DataMonitor.cs
using System;

namespace lab1a{
    public class DataMonitor{
        int capacity;
        Car[] cars;
        int lastIndex;
        public DataMonitor(int capacity = 32){
            this.capacity = capacity;
            cars = new Car[capacity];
            lastIndex = -1;
        }

        public bool AddCar(Car car){
            if(lastIndex == capacity){
                return false;
            }

            cars[++lastIndex] = car;

            return true;
        }

        public Car RemoveCar(){
            if(lastIndex == -1){
                return null;
            }
            else{
                return cars[lastIndex--];
            }
        }
    }
}

File: lab1_cpp/lab1_a/DataMonitor.cpp
#include "DataMonitor.h"
#include <iostream>

DataMonitor::DataMonitor(int cap)
{
    capacity = cap;
    cars = new car[capacity];
    count = 0;
    front = -1;
    rear = -1;
}

void DataMonitor::insert(car c)
{
    std::unique_lock<std::mutex> guard(mut);
	cv.wait(guard, [&]{ return count < capacity;});

  
    cars[count ++] = c;
    cv.notify_all();
}

car DataMonitor::remove()
{
    std::unique_lock<std::mutex> guard(mut);
    cv.wait(guard, [&]{ return count > 0; });

    count --;
    cv.notify_all();
    return cars[count];
}

bool DataMonitor::is_empty()
{
    std::unique_lock<std::mutex> guard(mut);
    return count == 0;
}

void DataMonitor::insert_first(car c)
{
    std::unique_lock<std::mutex> guard(mut);
	cv.wait(guard, [&]{ return count < capacity;});
    count ++;
    for(int i = count - 1; i >= 1; i --)
    {
        cars[i] = cars[i - 1];
    }
    cars[0] = c;
   
    cv.notify_all();
}
int DataMonitor::size(){
    return count;
}